### DISCUSSION TITLE
http vs https

### DISCUSSION THREAD

**Discussion by JUDONGHYEOK** [woowacourse-study/2022-woowahan-bansanghwe#101] [Network]
HTTPS는 HTTP의 가장 유명한 보안버전이다. HTTPS는 모든 HTTP 요청과 응답 데이터는 네트워크로 보내지기 전에 암호화 된다. HTTPS는 전송계층으로 넘어가기전 SSL 혹은 TLS로 이루어진 보안 계층을 거쳐 전송계층으로 넘어가기 때문에 HTTP보다 보안측면에서 유리하다.

### HTTPS 절차

1. HTTPS에서 클라이언트는 먼저 웹서버의 443 포트로 연결한다.
2. TCP가 연결되고 나면, 클라이언트와 서버는 암호법 매개변수와 교환키를 협상하면서 SSL 계층을 초기화한다.
3. 핸드셰이크가 완료되면 SSL 초기화는 완료되며, 클라이언트는 요청 메시지를 보안 계층에 보낼 수 있다.
4. 이 메시지는 TCP로 보내지기 전에 암호화된다.

### 인증서 발급 절차

1. A기업은 HTTP 기반의 애플리케이션에 HTTPS를 적용하기 위해 공개키/개인키를 발급함
2. CA 기업에게 돈을 지불하고, 공개키를 저장하는 인증서의 발급을 요청함
3. CA 기업은 CA기업의 이름, 서버의 공개키, 서버의 정보 등을 기반으로 인증서를 생성하고, CA 기업의 개인키로 암호화하여 A기업에게 이를 제공함
4. A기업은 클라이언트에게 암호화된 인증서를 제공함
5. 브라우저는 CA기업의 공개키를 미리 다운받아 갖고 있어, 암호화된 인증서를 복호화함
6. 암호화된 인증서를 복호화하여 얻은 A기업의 공개키로 세션키를 공유함

![image](https://user-images.githubusercontent.com/65863017/192131020-f884484a-21bf-47b7-97d7-e543e2b0509d.png)

### SSL 핸드셰이크

암호화된 HTTP 메시지를 보낼 수 있게 되기 전에, 클라이언트와 서버는 SSL 핸드셰이크를 할 필요가 있다. 

- 프로토콜 버전 번호 교환
- 양쪽이 알고 있는 암호 선택
- 양쪽의 신원 인증
- 채널을 암호화하기 위한 임시 세션 키 설정

1. 클라이언트가 암호 후보들을 보내고 인증서를 요구한다.
2. 서버는 선택된 암호와 인증서를 보낸다.(인증서에 서버에 공개키 포함)
3. 클라이언트가 비밀정보를 보낸다. 클라이언트와 서버는 키를 만든다.
4. 클라이언트와 서버는 서로에게 암호화를 시작한다고 말해준다.

보안 HTTPS 트랜잭션은 항상 서버 인증서를 요구한다. 잘 알려진 인증기관에 의해 서명된 서버 인증서는 서버에 개인정보를 보내기 전에 그 서브를 얼마나 신뢰할 수 있는지 평가하는 것을 도와줄 것이다.

최신 웹브라우저들 대부분은 인증서에 대해 간단하게 기본적인 검사를 하고 그 결과를 더 철저한 검사를 할 수 있는 방법과 함께 유저에게 제공한다. 

### HTTPS를 사용하는 이유

1. 통신의 기밀성
    1. 클라이언트와 서버의 통신을 보호한다.
2. 데이터의 무결성
    1. 변조되지 않은 데이터를 목적지에 도달하게 한다.
3. 인증
    1. HTTPS의 인증서를 통해 믿을만한 사이트인지 검증할 수 있음

2014년 구글에선 https를 사용한 사이트가 SEO에서도 가산점을 부과받을 수 있다고 권고하였다.

**Comment by kamwoo** [comment]
https에 대해서 설명해주세요

**Comment by kamwoo** [comment]
ssl이란 무엇인가요?

**Comment by kamwoo** [comment]
ssl 핸드세이크에 대해서 설명해주세요

**Comment by kamwoo** [comment]
https의 연결절차를 설명해주세요

**Comment by kamwoo** [comment]
http란 무엇인가요?

**Comment by kamwoo** [comment]
http 요청과 응답은 어떻게 이루어지나요?

**Comment by kamwoo** [comment]
http 연결 후에는 어떻게 전송하나요?

**Comment by kamwoo** [comment]
전송계층에서 TCP와 UDP의 공통점과 차이점을 설명해주세요

**Comment by kamwoo** [comment]
UDP를 언제 이용하나요?

**Comment by kamwoo** [comment]
HTTP 1.0 과 1.1의 차이점

**Comment by kamwoo** [comment]
HTTP 2는 어떻게 달라졌나요?

**Comment by kamwoo** [comment]
대칭키에 대해서 설명해주세요

**Comment by kamwoo** [comment]
4way 핸드세이크에 대해서 설명해주세요
